### PR TITLE
MultilineText Menu

### PR DESCRIPTION
**Pull request by chien**
Changes proposed in this pull request:
- Support MultilineText menu
- Modify Example to show how PagingMenuController works with other views

### CODE REVIEW

File: PagingMenuController.podspec [hunk 1]
```diff
@@ -9,19 +9,15 @@
 
 Pod::Spec.new do |s|
   s.name             = "PagingMenuController"
-  s.version          = "0.9.1"
+  s.version          = "0.9.2"
   s.summary          = "A paging view with customizable menu"
-  s.homepage         = "https://github.com/kitasuke/PagingMenuController"
-  # s.screenshots     = "www.example.com/screenshots_1", "www.example.com/screenshots_2"
+  s.homepage         = "https://github.com/chien/PagingMenuController"
```
Comment (kitasuke) [inline PagingMenuController.podspec]: Why do we have to use your repository?

Comment (chien) [inline PagingMenuController.podspec]: My bad. Will change it back

On Mon, May 16, 2016 at 00:44, Yusuke Kita <notifications@github.com> wrote:
In PagingMenuController.podspec [https://github.com/kitasuke/PagingMenuController/pull/207#discussion_r63309487] :

>    s.summary          = "A paging view with customizable menu"
> -  s.homepage         = "https://github.com/kitasuke/PagingMenuController"
> -  # s.screenshots     = "www.example.com/screenshots_1", "www.example.com/screenshots_2"
> +  s.homepage         = "https://github.com/chien/PagingMenuController"


Why do we have to use your repository?

**Comment by chien** [comment]
Podspec Updated

**Comment by chien** [comment]
I think this pull request also solves #153. Let me know if you want me to change anything.

File: Pod/Classes/MenuView.swift [hunk 2]
```diff
@@ -85,21 +85,21 @@ public class MenuView: UIScrollView {
     
     // MARK: - Lifecycle
     
-    internal init(menuItemTitles: [String], menuItemDescs: [String], options: PagingMenuOptions) {
+    public init(menuItemTitles: [String], menuItemDescs: [String], options: PagingMenuOptions) {
```
Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: Even if you make it public, I don't think MenuView's layout works properly.
I'm working on this with https://github.com/kitasuke/PagingMenuController/pull/197.
Is this change relevant to multiline text menu?

**Comment by kitasuke** [comment]
I accidentally commented on an outdated diff... Check and see my latest comment.

File: Pod/Classes/MenuView.swift (hunk 2 above)
Comment (chien) [inline Pod/Classes/MenuView.swift]: It works as static header for me. I use it in my project.
menuItemCount needs to be specifically set for now. 



```
let pageMenuOption: PagingMenuOptions = {
     let options = PagingMenuOptions()
     options.menuItemCount = 3
     return options
}()
displayInfoView = MenuView(menuItemTitles: ["title 1", "title 2", "title 3"], options: options)
self.addSubview(displayInfoView!)
```

Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: To be more specific, does it display properly when screen is rotated? Or does it recognize tap gesture?

Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: As long as it's a static header, I think its fine. However I'm not sure everyone use it in a same way. So could you revert this change if its not relevant to multiline text mode?
I'm taking care of this feature.

Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: There should be more explanations for this usage. Well, it works fine for now. However, a new feature which is separable menu view will be launched soon. It won't work with the feature in this way then.

Comment (chien) [inline Pod/Classes/MenuView.swift]: Maybe I should wait till separable menu view is merged back to master? 
I can add the multiline menu support later. It works for my project in my own repo for now. If you think this feature would be good, I'm happy to rework on the multiline text thing with the new code of separable menu view later.

Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: It should be done soon, but I can merge the multiline text part only first!
Well, its safer for you to wait the feature though.
Sorry for my inconvenience.

Comment (chien) [inline Pod/Classes/MenuView.swift]: I'll wait for the feature then. No need to be sorry. Thanks for maitaining this project. Save me a lot of time. :)

On Mon, May 16, 2016 at 01:35, Yusuke Kita <notifications@github.com> wrote:
In Pod/Classes/MenuView.swift [https://github.com/kitasuke/PagingMenuController/pull/207#discussion_r63311602] :

> @@ -85,21 +85,21 @@ public class MenuView: UIScrollView {
>      
>      // MARK: - Lifecycle
>      
> -    internal init(menuItemTitles: [String], menuItemDescs: [String], options: PagingMenuOptions) {
> +    public init(menuItemTitles: [String], menuItemDescs: [String], options: PagingMenuOptions) {


It should be done soon, but I can merge the multiline text part only first!
Well, its safer for you to wait the feature though.
Sorry for my inconvenience.

**Comment by chien** [comment]
@kitasuke updated with the latest master.

**Comment by kitasuke** [comment]
Thanks for your quick response!
Can you run example project? I've tried both of them, but neither of them didn't work...

**Comment by chien** [comment]
Updated pod spec. Can you try again?

**Comment by kitasuke** [comment]
Thanks! Can you also check what's going on in PagingMenuControllerDemo as well? It gets fatal error.

**Comment by chien** [comment]
It's working for me though. both PagingMenuControllerDemo and PagingMenuControllerDemo2

**Comment by MarcelStructr** [comment]
Hello,
thank you for your work @chien and @kitasuke. I am trying to use the version of @chien right now to display a MenuView without any Controller in my code. 

I was using this in viewDidLoad():

        // Do any additional setup after loading the view, typically from a nib.
        
        let options = PagingMenuOptions()

        options.menuItemMargin = 5
        options.menuHeight = 40
        options.menuDisplayMode = .SegmentedControl
        options.menuItemViewContent = .MultilineText
        
        let menu = MenuView(menuItemTypes: ["Category1", "Category2", "Category3", "Category4", "Category5"], options: options)
        
        view.addSubview(menu)

But I get a fatal error, can you help me out? I guess I have to modify the option-settings.
Example1 and 2 don't seem exactly that, what I want, so I wanted to ask you for your help.

**Comment by kitasuke** [comment]
If you want to use menu view alone, follow code below.
It's my mistake to make MenuView initialization method public. It should be internal.

```
        let options = PagingMenuOptions()
        options.menuDisplayMode = .SegmentedControl
        options.menuComponentType = .MenuView

        let pagingMenuController = PagingMenuController(menuControllerTypes: ["Category1", "Category2"], options: options)
        
        addChildViewController(pagingMenuController)
        view.addSubview(pagingMenuController.view)
        pagingMenuController.didMoveToParentViewController(self)
```

**Comment by MarcelStructr** [comment]
This is weird, i am getting following error in the library. I still use the repository of @chien :

http://www2.pic-upload.de/img/30686665/error.jpg

Okay, all right I can also wait (:

**Comment by kitasuke** [comment]
if you just want to use menu view alone, use master branch. It's already merged!

**Comment by MarcelStructr** [comment]
So by using Cocoapods in my project it seems, that it's still an old version. 
After that I manually downloaded the master branch and added the files into my project. I get the same error then: http://www2.pic-upload.de/img/30686665/error.jpg 

Sorry, for the inconvenience!

**Comment by chien** [comment]
@kitasuke Are you able to run example project using this branch after cocoapod upgrade?

**Comment by kitasuke** [comment]
I'm not able to run the project before cocoapods update, so its still not working for me.

**Comment by chien** [comment]
I've updated the branch with cocoa pod 1.0 update you pushed to the master. Can you check the branch again and let me know?
Thanks a lot.

On Tue, May 17, 2016 at 21:06, Yusuke Kita <notifications@github.com> wrote:
I'm not able to run the project before cocoapods update, so its still not working for me.

**Comment by kitasuke** [comment]
No its still not working... I'll look into more details.

**Comment by chien** [comment]
Fixed the crash for iPad. Not sure if it's the same problem you had though.

**Comment by kitasuke** [comment]
Thanks for merging master branch, but I don't see any references of some files you added in xcode project. I can see them in `PagingMenuControllerDemo2` folder though.

**Comment by chien** [comment]
@kitasuke references added.

**Comment by kitasuke** [comment]
Thanks, but I still get fatal error...
Can you re-create a new PR with this change? Or if it bothers you, I can do that instead.
Which way do you prefer?

**Comment by kitasuke** [comment]
It might be because of `PagingMenuOptions` object. I'm looking into more detail anyway.

**Comment by chien** [comment]
Not sure what happened ... 
How did you get the fatal error? By running the demo projects?
And how errors did you get?

**Comment by kitasuke** [comment]
When a new option is added into `PagingMenuOptions` class, I get the error below.
It's run time error.
I guess there is nothing wrong with your PR, but give me some time to investigate more.

screen shot 2016-05-19 at 21 05 49

**Comment by chien** [comment]
if you are getting reference error, Clean Build Folder solved the problem for me.

**Comment by kitasuke** [comment]
It's not reference error, but I found out the reason!

File: Pod/Classes/MenuItemView.swift [hunk 3]
```diff
@@ -18,6 +18,16 @@ public class MenuItemView: UIView {
...
         return label
     }()
+    
+    lazy public var descLabel: UILabel = {
+        let label = UILabel(frame: .zero)
+        label.numberOfLines = 1
+        label.textAlignment = .Center
+        label.userInteractionEnabled = true
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
```
Comment (kitasuke) [inline Pod/Classes/MenuItemView.swift]: Can you make a common method for initializing UILabel?
It's redundant since its exactly same code with `titleLabel.`

File: Pod/Classes/MenuItemView.swift [hunk 4]
```diff
@@ -145,6 +178,18 @@ public class MenuItemView: UIView {
         translatesAutoresizingMaskIntoConstraints = false
     }
     
+    private func setupLabels(title: String, desc: String) {
+        titleLabel.text = title
+        titleLabel.textColor = options.textColor
+        titleLabel.font = options.font
+        addSubview(titleLabel)
```
Comment (kitasuke) [inline Pod/Classes/MenuItemView.swift]: Make `setupTitleLabel` and `setupDescriptionLabel`, then you can call `setupTitleLabel` in `setupDescriptionLabel`.

File: Pod/Classes/MenuItemView.swift [hunk 5]
```diff
@@ -162,6 +207,28 @@ public class MenuItemView: UIView {
...
+        let verticalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("V:|-[titleLabel(titleHeight)][descLabel(descHeight)]-|", options: [], metrics: metrics, views: viewsDictionary)
+        
+        let descHorizontalConstraints = NSLayoutConstraint.constraintsWithVisualFormat("H:|[descLabel]|", options: [], metrics: nil, views: viewsDictionary)
+        
+        NSLayoutConstraint.activateConstraints(horizontalConstraints + verticalConstraints + descHorizontalConstraints)
+        
+        widthConstraint = NSLayoutConstraint(item: titleLabel, attribute: .Width, relatedBy: .Equal, toItem: nil, attribute: .Width, multiplier: 1.0, constant: labelSize.width)
+        widthConstraint.active = true
+        widthConstraint = NSLayoutConstraint(item: descLabel, attribute: .Width, relatedBy: .Equal, toItem: nil, attribute: .Width, multiplier: 1.0, constant: labelSize.width)
+        widthConstraint.active = true
```
Comment (kitasuke) [inline Pod/Classes/MenuItemView.swift]: `widthConstraint` is for `titleLabel`. Two constraint properties for both `titleLabel` and `descriptionLabel` are needed to change width dynamically.

File: Pod/Classes/MenuView.swift [hunk 6]
```diff
@@ -8,6 +8,16 @@
 
 import UIKit
 
+public struct MultilineMenuItem: MenuItemType {
+    var title = ""
+    var desc = ""
```
Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: There is no point that `var` is used.
Please use `let`.

```
let title: String
let description: String
```

File: Pod/Classes/PagingMenuOptions.swift [hunk 7]
```diff
@@ -28,11 +32,12 @@ public class PagingMenuOptions {
     public var menuItemMode = MenuItemMode.Underline(height: 3, color: UIColor.blueColor(), horizontalPadding: 0, verticalPadding: 0)
     public var lazyLoadingPage: LazyLoadingPage = .Three
     public var menuControllerSet: MenuControllerSet = .Multiple
+    public var menuItemViewContent: MenuItemViewContent = .Text
```
Comment (kitasuke) [inline Pod/Classes/PagingMenuOptions.swift]: Please change to `internal`. 
It's automatically set inside `willSet`, so no need for user to explicitly set it. 
https://github.com/kitasuke/PagingMenuController/pull/207/files#diff-ddd20afa2d86943bffcfe84300a5cfcdR33

File: Pod/Classes/UIViewController+MenuItemImage.swift [hunk 8]
```diff
@@ -19,4 +21,29 @@ public extension UIViewController {
...
+        get {
+            guard let desc = objc_getAssociatedObject(self, &MenuItemDescKey) as? String else { return nil }
+            return desc
+        }
+        set {
+            objc_setAssociatedObject(self, &MenuItemDescKey, newValue, .OBJC_ASSOCIATION_RETAIN)
+        }
+    }
+    
+    var menuItemView: MenuItemView? {
```
Comment (kitasuke) [inline Pod/Classes/UIViewController+MenuItemImage.swift]: This way which sets custom property is not a good one. Therefore, it shouldn't be used unnecessary property like `menuItemView`. It can be accessed by `PagingMenuController` instance.
Please revert this code, plus `updateMenuContent` method.

**Comment by kitasuke** [comment]
- I got constraint error in PagingMenuController2. Please fix it.

```
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
2016-05-20 08:57:59.562 PagingMenuControllerDemo2[27950:896618] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x7fb90b7398b0 V:[UILabel:0x7fb90b72a910'Multi 3'(30)]>",
    "<NSLayoutConstraint:0x7fb90b7399b0 V:[UILabel:0x7fb90b72ab20'Desc'(24)]>",
    "<NSLayoutConstraint:0x7fb90b72cb80 UILabel:0x7fb90b72a910'Multi 3'.top == PagingMenuController.MenuItemView:0x7fb90b73f0c0.topMargin>",
    "<NSLayoutConstraint:0x7fb90b729f50 V:[UILabel:0x7fb90b72a910'Multi 3']-(0)-[UILabel:0x7fb90b72ab20'Desc']>",
    "<NSLayoutConstraint:0x7fb90b739a00 PagingMenuController.MenuItemView:0x7fb90b73f0c0.bottomMargin == UILabel:0x7fb90b72ab20'Desc'.bottom>",
    "<NSLayoutConstraint:0x7fb90b72b190 V:|-(0)-[PagingMenuController.MenuItemView:0x7fb90b73f0c0]   (Names: '|':UIView:0x7fb90b727760 )>",
    "<NSLayoutConstraint:0x7fb90b72b430 V:[PagingMenuController.MenuItemView:0x7fb90b73f0c0]-(0)-|   (Names: '|':UIView:0x7fb90b727760 )>",
    "<NSLayoutConstraint:0x7fb90b7295c0 V:[PagingMenuController.MenuView:0x7fb90c836600(60)]>",
    "<NSLayoutConstraint:0x7fb90b726a70 UIView:0x7fb90b727760.height == PagingMenuController.MenuView:0x7fb90c836600.height>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x7fb90b7398b0 V:[UILabel:0x7fb90b72a910'Multi 3'(30)]>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKit/UIView.h> may also be helpful.
```

- `description` is preferred, not `desc`.
- PagingMenuControllerDemo has all the examples with specific option now, not PagingMenuControllerDemo2. Could you move its usage to PagingMenuControllerDemo? Since this change was done after your PR, I can do this instead.

File: Pod/Classes/MenuView.swift [hunk 9]
```diff
@@ -8,6 +8,16 @@
 
 import UIKit
 
+public struct MultilineMenuItem: MenuItemType {
```
Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: There is no need to be public. Please make it internal.

Comment (chien) [inline Pod/Classes/MenuView.swift]: In the case of creating view only menu with MultilineMenuItem, we need to pass [MultilineMenuItem] when initializing the PagingMenuViewController

Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: oh I don't have any plan to provide view only for now.
I forgot to mention that.

File: Pod/Classes/MenuView.swift [hunk 10]
```diff
@@ -8,6 +8,16 @@
...
+    var title = ""
+    var desc = ""
+    
+    public init(title: String, desc: String) {
```
Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: Please make it internal.

File: Pod/Classes/MenuView.swift (hunk 9 above)
Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: At least, this PR is about multiline text menu. Please create a new PR of menu view only usage.
I'll consider it at that time.

Comment (chien) [inline Pod/Classes/MenuView.swift]: Not even for this?
```
PagingMenuController(menuItemTypes: ["Viewable Menu1", "Viewable Menu 2"], options: options)
```

if we make this internal, MultilineMenuItem would not work for menuItemTypes.

Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: Sorry I was wrong. You are right, public looks good!

**Comment by chien** [comment]
Thanks a lot for your comments. The changes are made except for two. Please let me know if my explanation makes sense to you.

The use case of MultilineMenuItem is to change the description of the menu based on the action taken on the view controller. And to be able to change the description of a menu item, I believe the following things need to happen:

1. MultilineMenuItem should have var instead of let
2. updateMenuContent will separate the concern of how this library updates the menu item view from application's view controller. Otherwise, it would need to know get the current menu item from paging menu view controller as below (at least that's how I understand the code). And this approach means application needs to know exactly how the library keeps all these references. And future changes of the library might break applications.

```
pagingMenuViewController.menuView.currentMenuItemView.desc = "..."
```

File: Pod/Classes/PagingMenuOptions.swift [hunk 11]
```diff
@@ -21,8 +21,8 @@ public class PagingMenuOptions {
     public var descTextColor = UIColor.lightGrayColor()
     public var menuPosition: MenuPosition = .Top
     public var menuTitleHeight: CGFloat = 30
-    public var menuDescHeight: CGFloat = 24
-    public var menuHeight: CGFloat = 54
+    public var menuDescHeight: CGFloat = 21
```
Comment (kitasuke) [inline Pod/Classes/PagingMenuOptions.swift]: Can you also change this property name to menuDescriptionHeight?

**Comment by kitasuke** [comment]
Thanks for your quick work!
I agree with your opinion. I'll accept your PR for now, but please understand it'll be changed soon. I'm rewriting this library in protocol oriented way to avoid dangerous usage like getting or setting associated object.

Just added two comments.

File: Pod/Classes/MenuView.swift (hunk 6 above)
Comment (kitasuke) [inline Pod/Classes/MenuView.swift]: then set type only here like below.

```
var title: String
var description: String
```

File: Pod/Classes/PagingMenuOptions.swift [hunk 12]
```diff
@@ -17,8 +17,12 @@ public class PagingMenuOptions {
     public var selectedTextColor = UIColor.blackColor()
     public var font = UIFont.systemFontOfSize(16)
     public var selectedFont = UIFont.systemFontOfSize(16)
+    public var descFont = UIFont.systemFontOfSize(12)
+    public var descTextColor = UIColor.lightGrayColor()
     public var menuPosition: MenuPosition = .Top
-    public var menuHeight: CGFloat = 50
+    public var menuTitleHeight: CGFloat = 30
+    public var menuDescHeight: CGFloat = 21
```
Comment (kitasuke) [inline Pod/Classes/PagingMenuOptions.swift]: Can you also change this property name to menuDescriptionHeight?

**Comment by chien** [comment]
protocol oriented way would be great!

Changes made. Please help take a look. Thanks a lot.

**Comment by kitasuke** [comment]
Well done!
I really appreciate your work :)
